common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java

```diff
@@ -110,48 +115,68 @@ public static void submitSilent(@SuppressWarnings(WarningType.rawtype_warning) @
-                commandResult = sCommandResultCache.get(command);
+                    CommandResult commandResult = null;
 
-                //If nothing in cache, execute the command and cache the result
-                if (commandResult == null) {
-                    commandResult = executeCommand(command);
-                    cacheCommandResult(command, commandResult);
-                } else {
-                    Logger.info(
-                            TAG + methodName,
-                            "Silent command result returned from cache."
-                    );
-                }
+                    //Log operation parameters
+                    if (command.getParameters() instanceof SilentTokenCommandParameters) {
+                        logSilentRequestParams(methodName, (SilentTokenCommandParameters) command.getParameters());
+                        EstsTelemetry.getInstance().emitForceRefresh(((SilentTokenCommandParameters) command.getParameters()).isForceRefresh());
+                    }
 
-                // set correlation id on Local Authentication Result
-                setCorrelationIdOnResult(commandResult, correlationId);
+                    //Check cache to see if the same command completed in the last 30 seconds
+                    commandResult = sCommandResultCache.get(command);
+
+                    //If nothing in cache, execute the command and cache the result
+                    if (commandResult == null) {
+                        commandResult = executeCommand(command);
+                        cacheCommandResult(command, commandResult);
+                    } else {
+                        Logger.info(
+                                TAG + methodName,
+                                "Silent command result returned from cache."
+                        );
+                    }
 
-                Telemetry.getInstance().flush(correlationId);
-                EstsTelemetry.getInstance().flush(command, commandResult);
+                    // set correlation id on Local Authentication Result
+                    setCorrelationIdOnResult(commandResult, correlationId);
 
-                //Return the result via the callback
-                sExecutingCommandMap.remove(command).setResult(commandResult);
+                    Telemetry.getInstance().flush(correlationId);
+                    EstsTelemetry.getInstance().flush(command, commandResult);
+                    finalFuture.setResult(commandResult);
+                    //Return the result via the callback
+                } catch (Throwable t) {
+                    finalFuture.setException(new ExecutionException(t));
+                } finally {
+                    final ResultFuture mapFuture = sExecutingCommandMap.remove(command);
+                    if (mapFuture == null) {
+                        // If this has happened, the command that we started with has mutated.  We will
```

Comment:
May I ask what you meant by 'mutated' here? In what (hypothetical) scenario this might happen?

Reply:
If the object used as the key in the map is changed in such a way to cause it no longer to have the same value via equals/hashCode, then it will not be found in the map when we go to remove it.

Reply:
(also makes me wonder if the command object itself it a good key here)

Reply:
We could generate a key from the command object and remove that worry, too.

Reply:
Did we find anything (so far) within the command that could change during execution?

Reply:
That might be better, but I think that'll be big and for now this fix should prevent NPE in the meantime. Let's put a todo here and create a tracking work item maybe?

Reply:
> Did we find anything (so far) within the command that could change during execution?

Nothing obvious.  If you look at the other CR, I put a bunch of safe-copy in place where I found it, but the scope is huge.  The thing with this is that it should be safe even if we write bugs into it, and that's really best.

Reply:
Yup, agreed!

Reply:
I'm actually going to wait on this until we have a unit test.  Brian and I are both trying to write some so there are more eyes on it.

Reply:
And if anybody else wants to play this game, more different viewpoints are better.

Reply:
Brian and I have added some tests.  I've put the todo item in place.